core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1716,7 +1739,31 @@ private[spark] class DAGScheduler(
             if (runningStages.contains(shuffleStage) && shuffleStage.pendingPartitions.isEmpty) {
               if (!shuffleStage.shuffleDep.shuffleMergeFinalized &&
                 shuffleStage.shuffleDep.getMergerLocs.nonEmpty) {
-                scheduleShuffleMergeFinalize(shuffleStage)
+                // Check if a finalize task has already been scheduled. This is to prevent the
+                // following scenario: Stage A attempt 0 fails and gets retried. Stage A attempt 1
+                // succeeded, triggering the scheduling of shuffle merge finalization. However,
+                // tasks from Stage A attempt 0 might still be running and sending task completion
+                // events to DAGScheduler. This check prevents multiple attempts to schedule merge
+                // finalization get triggered due to this.
+                if (shuffleStage.shuffleDep.getFinalizeTask.isEmpty) {
+                  // If total shuffle size is smaller than the threshold, attempt to immediately
+                  // schedule shuffle merge finalization and process map stage completion.
+                  val totalSize = Try(mapOutputTracker
+                    .getStatistics(shuffleStage.shuffleDep).bytesByPartitionId.sum).getOrElse(0L)
```

Comment:
This would fail when some of the tasks outputs are lost due to node/executor loss which basically causes `MapOutputTracker.getStatistics` throw exception. Thoughts?

Reply:
I do not see other uses of `getStatistics` catching exceptions ... hence my query.
Are we expecting NPE's here ?

Assuming exception and so 0L - we are immediately finalizing here : do we want to do that only for deterministic stage ?

Reply:
Other usages of `getStatistics` are with in `mapStage.isAvailable` check so its guaranteed to not throw exception. Unfortunately, that is not the case here.

Do you mean we can avoid finalizing the shuffle for non-deterministic stage in this case?

Reply:
That makes sense, thanks for clarifying about `isAvailable`.

Yes, if `!isAvailable`, avoid finalizing for non-deterministic stage ? (for deterministic stage, with push based shuffle enabled, we should finalize)

Reply:
I think we should `finalize` even `non-deterministic stage` + `!isAvailable`, if not then those file handles created for merging shuffle files won't get closed. This is because in the retry we are refreshing merger locations. But what we can do is `finalize` the shuffle for `non-deterministic stage` + `!isAvailable` with `registerMergeResults` set to false that way it asynchronously finalizes with all shuffle services but we would mark the shuffle merge finalized and won't keep DAGScheduler waiting. Thoughts?

Reply:
Need to finalize for all stages, irrespective of  failed or not, makes sense - thanks for clarifying !

For non-deterministic stage, we should do what is being done right now - immediately finalize and ignore all ESS response when stage is not available (since we are not going to use it).

What we need to address here is that we are throwing away deterministic stage output due to some output being unavailable - which can actually be reused during retry.

This is a consequence of using `getStatistics` api (NPE is thrown from `getStatistics` and `Try` will return 0L).
This needs to be addressed.

A rough sketch would be, we mirror functionality from `ShufflePartitionUtil.getMapSizesForReduceId` and set size to `-1` when `MapStatus` is null.

```
if (shuffleStage.shuffleDep.getFinalizeTask.isEmpty) {

  // Comment:
  // if failed and NON_DET - immediately fail.
  // if failed and DET - decide based on size available and shuffleMergeWaitMinSizeThreshold
  // If available, decide based on size available and shuffleMergeWaitMinSizeThreshold

  val totalSize = {
    lazy val computedTotalSize = MOT.getStatistics(). bytesByPartitionId .filter(_ > 0).sum()
    // note: isAvialble is cheap compared to computing it via getStatistics output for missing
    if (shuffleStage.isAvailable) {
      computedTotalSize
    } else {
      if (shuffleStage.isNonDeterministic) {
        0L
      } else {
        // Modify `getStatistics` to be null aware for `MapStatus` .
        computedTotalSize
      }
    }
  }
}

```

Thoughts ?